dependent_types.md

```diff
@@ -0,0 +1,279 @@
+```
+{-# LANGUAGE DataKinds               #-}
+{-# LANGUAGE KindSignatures          #-}
+{-# LANGUAGE MultiParamTypeClasses   #-}
+{-# LANGUAGE FunctionalDependencies  #-}
+{-# LANGUAGE FlexibleInstances       #-}
+module Main where
+
+import GHC.TypeLits
+
+data Label (l :: Symbol) = Get
+data Point = Point Int Int deriving Show
+
+class Has a l b | a l -> b where
+  from :: a -> Label l -> b
+
+instance Has Point "x" Int
+    where from (Point x _) _ = x
+
+instance Has Point "y" Int
+    where from (Point _ y) _ = y
+
+example :: Int
+example = from (Point 1 2) (Get :: Label "x")
+
+```
+So type level data is something that lives on the level of types but is not a type.
+
+## Lambda Cube
+
+[Lambda Cube](https://en.wikipedia.org/wiki/Lambda_cube) is the term from Type Theory and Math Logic. It builds on top of [lambda calculus](https://en.wikipedia.org/wiki/Lambda_calculus) defining all possible type systems where each axes is the new abstraction. So going in some order from simple to more complex systems we first arrive to what every programming language has :
+
+**Values depending on values** (_functions_)
+- This simply means that you have some sort of relation between variables if they are used in terms of each other. 
+
+Example:
+
+```
+a :: Integer
+b = even a :: Bool
+-- here b depends on a 
+```
+
+After that we get to the second step which is
+
+**Values depending on Types** (_classes_)
+
+Example:
+
+```
+max :: Integer
+max :: Double
+```
+
+We can look at this as if type is determining the values which constant `max` can hold. This is a basic relation between values and types, values need to be of some type in Haskell.
+
+Third step in our wonderfull type land is 
+
+**Types depending on Types** (_type functions_)
+
+In order to demonstrate what we mean by this we will use Type Families. These [slides](https://cdepillabout.github.io/haskell-type-families-presentation/#/) do really good job in explaining what are type families.
+
+Basically they provide a way to define a function that operates on types. Depending on the concrete implementation we can return different types for different instances of the typeclass or at least we have that choice.
+
+This leads us to the final step to dependent types which is: 
+
+**Types depending on Values** (_dependent types_)
+
+As I mentioned at the beginning Haskell still does not have native support for dependent types but we have a handfull of extensions that provide a way to touch on that. We are now arriving to the next step in understanding dependent types which is
+
+## Local Assumptions
+
+GADTs or Generalized Algebraic DataTypes is the Haskell extension that provides us with _local assumptions_. What does that mean? 
+
+Well, let us look at some of the standard Haskell datatypes :
+
+```
+data Maybe a = Just a | Nothing
+data Either a b = Left a | Right b
+```
+If we rewrote these types with hypotetical syntax we would come to this:
+```
+data Maybe a = Just a
```

Comment:
If the article is aimed at beginners, here's an observation from personal experience. Beginners struggle to understand syntax that Haskell already has, and adding *more* syntax which is even not supported by the compiler (and is only hypothetical) does not aid understanding at all.

Reply:
Hypothetical syntax implies that a person can grok the higher-level idea behind it. I've seen a few Haskell Wiki introduce hypothetical syntax, and I dislike it a lot.